File: src/deephaven_mcp/resource_manager/_instance_tracker.py
"""
Instance tracking and orphaned resource cleanup for MCP server instances.

This module provides functionality to track running MCP server instances and clean up
orphaned Docker containers and python processes that may be left behind when a server
is terminated with SIGKILL or crashes unexpectedly.

Key Concepts:
    - Each MCP server instance is assigned a unique UUID on startup
    - Instance metadata (UUID, PID, start time, python processes) is persisted to disk
    - Docker containers are labeled with the instance UUID for identification
    - Python processes are tracked in the instance metadata file
    - On startup, dead instances are detected and their orphaned resources are cleaned up

Architecture:
    - Instance metadata stored in: ~/.deephaven-mcp/instances/{uuid}.json
    - Docker containers labeled with: deephaven-mcp-server-instance={uuid}
    - Python processes tracked in instance file: {"python_processes": {"session": pid}}

Usage:
    # On server startup
    instance = await InstanceTracker.create_and_register()
    await cleanup_orphaned_resources()

    # During operation
    await instance.track_python_process("my-session", 12345)
    await instance.untrack_python_process("my-session")

    # On server shutdown
    await instance.unregister()

Thread Safety:
    - All operations are coroutine-safe
    - File operations use atomic writes where possible
    - Multiple server instances can safely coexist
"""

import asyncio
import json
import logging
import os
import signal
import uuid
from datetime import datetime
from pathlib import Path

_LOGGER = logging.getLogger(__name__)


class InstanceTracker:
    """
    Tracks a single MCP server instance and its associated resources.

    This class manages the lifecycle of an MCP server instance, including:
    - Generating and persisting a unique instance identifier
    - Tracking python-launched session processes
    - Registering/unregistering the instance on startup/shutdown
    - Providing the instance ID for Docker container labeling

    Attributes:
        instance_id (str): Unique UUID for this server instance.
        pid (int): Process ID of this server instance.
        started_at (str): ISO 8601 timestamp of when this instance started.
        instance_file (Path): Path to the instance metadata file.
    """

    def __init__(self, instance_id: str, pid: int, started_at: str):
        """
        Initialize an InstanceTracker.

        This constructor should not be called directly. Use create_and_register()
        or load_from_file() factory methods instead.

        Args:
            instance_id (str): Unique UUID for this instance.
            pid (int): Process ID of this server instance.
            started_at (str): ISO 8601 timestamp of when the instance started.
        """
        self.instance_id = instance_id
        self.pid = pid
        self.started_at = started_at
        self._python_processes: dict[str, int] = {}

        # Ensure instances directory exists
        instances_dir = Path.home() / ".deephaven-mcp" / "instances"
        instances_dir.mkdir(parents=True, exist_ok=True)

        self.instance_file = instances_dir / f"{instance_id}.json"

    @classmethod
    async def create_and_register(cls) -> "InstanceTracker":
        """
        Create a new instance tracker and register it.

        This factory method creates a new instance with a unique UUID and immediately
        persists it to disk. Call this on MCP server startup.

        Returns:
            InstanceTracker: A new registered instance tracker.

        Example:
            ```python
            instance = await InstanceTracker.create_and_register()
            _LOGGER.info(f"Server instance: {instance.instance_id}")
            ```
        """
        instance_id = str(uuid.uuid4())
        pid = os.getpid()
        started_at = datetime.now().isoformat()

        tracker = cls(instance_id, pid, started_at)
        await tracker._save()

        _LOGGER.info(
            f"[InstanceTracker] Registered new instance {instance_id} (PID: {pid})"
        )

        return tracker

    @classmethod
    def load_from_file(cls, instance_file: Path) -> "InstanceTracker":
        """
        Load an existing instance tracker from a metadata file.

        This is used during orphan cleanup to load information about other
        (potentially dead) server instances.

        Args:
            instance_file (Path): Path to the instance metadata JSON file.

        Returns:
            InstanceTracker: Instance tracker loaded from the file.

        Raises:
            FileNotFoundError: If the instance file does not exist.
            json.JSONDecodeError: If the file contains invalid JSON.
            KeyError: If required fields (instance_id, pid, started_at) are missing from the JSON.
        """
        data = json.loads(instance_file.read_text())

        tracker = cls(
            instance_id=data["instance_id"],
            pid=data["pid"],
            started_at=data["started_at"],
        )
        tracker._python_processes = data.get("python_processes", {})

        return tracker

    async def track_python_process(self, session_name: str, pid: int) -> None:
        """
        Track a new python-launched session process.

        Adds the process to the instance metadata so it can be cleaned up
        if the server crashes or is killed.

        Args:
            session_name (str): Name of the session.
            pid (int): Process ID of the python-launched deephaven-server process.

        Example:
            ```python
            # After launching a python session
            await instance.track_python_process("my-session", 12345)
            ```
        """
        self._python_processes[session_name] = pid
        await self._save()

        _LOGGER.debug(
            f"[InstanceTracker] Tracking python process for session '{session_name}' (PID: {pid})"
        )

    async def untrack_python_process(self, session_name: str) -> None:
        """
        Stop tracking a python-launched session process.

        Removes the process from the instance metadata, typically called when
        the session is stopped normally.

        Args:
            session_name (str): Name of the session to stop tracking.

        Example:
            ```python
            # After stopping a python session
            await instance.untrack_python_process("my-session")
            ```
        """
        if session_name in self._python_processes:
            del self._python_processes[session_name]
            await self._save()

            _LOGGER.debug(
                f"[InstanceTracker] Stopped tracking python process for session '{session_name}'"
            )

    async def unregister(self) -> None:
        """
        Unregister this instance and remove its metadata file.

        Call this on normal server shutdown to clean up instance tracking.
        This prevents the cleanup logic from attempting to clean up resources
        for a server that shut down normally.

        Example:
            ```python
            # In app_lifespan finally block
            await instance.unregister()
            ```
        """
        try:
            self.instance_file.unlink(missing_ok=True)
            _LOGGER.info(f"[InstanceTracker] Unregistered instance {self.instance_id}")
        except Exception as e:
            _LOGGER.warning(
                f"[InstanceTracker] Error unregistering instance {self.instance_id}: {e}"
            )

    async def _save(self) -> None:
        """
        Save instance metadata to disk using atomic write.

        Persists the current state of the instance tracker, including tracked
        python processes, to the instance metadata file. Uses atomic write
        (temp file + rename) to ensure the file is never left in a corrupted
        state if the write is interrupted.
        """
        data = {
            "instance_id": self.instance_id,
            "pid": self.pid,
            "started_at": self.started_at,
            "python_processes": self._python_processes,
        }

        # Atomic write using temporary file + rename
        temp_file = self.instance_file.with_suffix(".tmp")
        temp_file.write_text(json.dumps(data, indent=2))
        temp_file.replace(self.instance_file)


def is_process_running(pid: int) -> bool:
    """
    Check if a process with the given PID is currently running.

    Uses os.kill(pid, 0) which sends signal 0 to check process existence
    without actually sending a signal. This is the standard Unix way to
    check if a process is alive.

    Args:
        pid (int): Process ID to check.

    Returns:
        bool: True if the process is running, False otherwise.

    Note:
        Returns False if the process doesn't exist OR if we don't have
        permission to check its status (e.g., process owned by another user).
    """
    try:
        os.kill(pid, 0)
        return True
    except OSError:
        return False


async def cleanup_orphaned_resources() -> None:
    """
    Clean up orphaned Docker containers and python processes from dead server instances.

    This function should be called on MCP server startup. It:
    1. Scans the instances directory for registered server instances
    2. Checks if each instance's process is still running
    3. For dead instances, cleans up their orphaned resources:
       - Stops and removes Docker containers (via instance label)
       - Kills python processes (from instance metadata)
    4. Removes instance metadata files for dead instances

    The cleanup is safe for concurrent server instances - only resources from
    dead servers are cleaned up. Running servers are left untouched.

    This handles the SIGKILL case where a server is forcibly terminated without
    a chance to clean up its resources in the finally block.

    Example:
        ```python
        # In app_lifespan, before yielding context
        await cleanup_orphaned_resources()
        ```

    Note:
        Errors during cleanup are logged but don't raise exceptions, ensuring
        that server startup continues even if cleanup partially fails.
    """
    instances_dir = Path.home() / ".deephaven-mcp" / "instances"

    if not instances_dir.exists():
        _LOGGER.debug(
            "[InstanceTracker] No instances directory, skipping orphan cleanup"
        )
        return

    instance_files = list(instances_dir.glob("*.json"))

    if not instance_files:
        _LOGGER.debug(
            "[InstanceTracker] No instance files found, skipping orphan cleanup"
        )
        return

    _LOGGER.info(
        f"[InstanceTracker] Checking {len(instance_files)} instance(s) for orphaned resources..."
    )

    for instance_file in instance_files:
        try:
            tracker = InstanceTracker.load_from_file(instance_file)

            # Check if this instance's process is still running
            if is_process_running(tracker.pid):
                _LOGGER.debug(
                    f"[InstanceTracker] Instance {tracker.instance_id} still running (PID {tracker.pid}), skipping"
                )
                continue

            # Process is dead - clean up its orphaned resources
            _LOGGER.warning(
                f"[InstanceTracker] Found dead instance {tracker.instance_id} (PID {tracker.pid}), cleaning up orphans..."
            )

            # Clean up Docker containers
            await _cleanup_docker_containers_for_instance(tracker.instance_id)

            # Clean up python processes
            await _cleanup_python_processes_for_instance(tracker)

            # Remove instance metadata file
            instance_file.unlink(missing_ok=True)
            _LOGGER.info(
                f"[InstanceTracker] Cleaned up orphaned resources for instance {tracker.instance_id}"
            )

        except Exception as e:
            _LOGGER.error(
                f"[InstanceTracker] Error cleaning up instance {instance_file.name}: {e}",
                exc_info=True,
            )


async def _cleanup_docker_containers_for_instance(instance_id: str) -> None:
    """
    Clean up Docker containers for a specific server instance.

    Finds all Docker containers labeled with the instance ID and stops/removes them.
    Uses the 'deephaven-mcp-server-instance' label to identify containers belonging
    to the dead instance. Attempts graceful stop via 'docker stop' before removing
    with 'docker rm'.

    Args:
        instance_id (str): The instance UUID to clean up containers for.

    Note:
        Errors during cleanup are logged but do not raise exceptions. This ensures
        that failure to clean up one container doesn't prevent cleanup of others
        or block server startup.
    """
    try:
        # Find containers with this instance ID label
        process = await asyncio.create_subprocess_exec(
            "docker",
            "ps",
            "-a",
            "--filter",
            f"label=deephaven-mcp-server-instance={instance_id}",
            "--format",
            "{{.ID}}",
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )

        stdout, stderr = await process.communicate()

        if process.returncode != 0:
            _LOGGER.warning(
                f"[InstanceTracker] Docker ps command failed: {stderr.decode()}"
            )
            return

        container_ids = [
            cid.strip() for cid in stdout.decode().strip().split("\n") if cid.strip()
        ]

        if not container_ids:
            _LOGGER.debug(
                f"[InstanceTracker] No Docker containers found for instance {instance_id}"
            )
            return

        _LOGGER.info(
            f"[InstanceTracker] Found {len(container_ids)} orphaned container(s) for instance {instance_id}"
        )

        # Stop and remove each container
        for container_id in container_ids:
            _LOGGER.info(
                f"[InstanceTracker] Stopping orphaned container {container_id[:12]}..."
            )

            # Try to stop gracefully
            stop_process = await asyncio.create_subprocess_exec(
                "docker",
                "stop",
                container_id,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            await stop_process.communicate()

            # Remove the container
            rm_process = await asyncio.create_subprocess_exec(
                "docker",
                "rm",
                container_id,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            await rm_process.communicate()

            _LOGGER.info(
                f"[InstanceTracker] Cleaned up orphaned container {container_id[:12]}"
            )

    except Exception as e:
        _LOGGER.error(
            f"[InstanceTracker] Error cleaning up Docker containers for instance {instance_id}: {e}",
            exc_info=True,
        )


async def _cleanup_python_processes_for_instance(tracker: InstanceTracker) -> None:
    """
    Clean up python processes for a specific server instance.

    Terminates all python processes tracked in the instance metadata. Sends SIGTERM
    to each tracked process after verifying it's still running. Processes that
    are already dead are logged and skipped.

    Args:
        tracker (InstanceTracker): The instance tracker with python process information.

    Note:
        Errors during process termination (e.g., permission denied, process already
        exited) are logged as warnings but do not raise exceptions. This ensures
        that cleanup continues for all processes even if some fail.
    """
    python_processes = tracker._python_processes

    if not python_processes:
        _LOGGER.debug(
            f"[InstanceTracker] No python processes found for instance {tracker.instance_id}"
        )
        return

    _LOGGER.info(
        f"[InstanceTracker] Found {len(python_processes)} orphaned python process(es) for instance {tracker.instance_id}"
    )

    for session_name, pid in python_processes.items():
        try:
            if is_process_running(pid):
                _LOGGER.info(
                    f"[InstanceTracker] Killing orphaned python process {pid} (session: {session_name})"
                )
                os.kill(pid, signal.SIGTERM)
            else:
                _LOGGER.debug(
                    f"[InstanceTracker] Python process {pid} (session: {session_name}) already dead"
                )
        except Exception as e:
            _LOGGER.warning(
                f"[InstanceTracker] Error killing python process {pid} (session: {session_name}): {e}"
            )

File: src-ts/resource-manager/instance-tracker.ts
/**
 * Instance tracking and orphaned resource cleanup for MCP server instances.
 *
 * This module provides functionality to track running MCP server instances and clean up
 * orphaned Docker containers and python processes that may be left behind when a server
 * is terminated with SIGKILL or crashes unexpectedly.
 *
 * Key Concepts:
 *   - Each MCP server instance is assigned a unique UUID on startup
 *   - Instance metadata (UUID, PID, start time, python processes) is persisted to disk
 *   - Docker containers are labeled with the instance UUID for identification
 *   - Python processes are tracked in the instance metadata file
 *   - On startup, dead instances are detected and their orphaned resources are cleaned up
 *
 * Architecture:
 *   - Instance metadata stored in: ~/.deephaven-mcp/instances/{uuid}.json
 *   - Docker containers labeled with: deephaven-mcp-server-instance={uuid}
 *   - Python processes tracked in instance file: {"python_processes": {"session": pid}}
 *
 * Usage:
 * ```typescript
 * // On server startup
 * const instance = await InstanceTracker.createAndRegister();
 * await cleanupOrphanedResources();
 *
 * // During operation
 * await instance.trackPythonProcess("my-session", 12345);
 * await instance.untrackPythonProcess("my-session");
 *
 * // On server shutdown
 * await instance.unregister();
 * ```
 *
 * Thread Safety:
 *   - All operations are async-safe
 *   - File operations use atomic writes where possible
 *   - Multiple server instances can safely coexist
 */

import * as fs from "node:fs/promises";
import * as path from "node:path";
import * as os from "node:os";
import { randomUUID } from "node:crypto";
import { spawn } from "node:child_process";
import pino from "pino";

const _logger = pino({ name: "deephaven-mcp:resource-manager/instance-tracker" });

// ---------------------------------------------------------------------------
// InstanceTracker
// ---------------------------------------------------------------------------

/**
 * Tracks a single MCP server instance and its associated resources.
 *
 * This class manages the lifecycle of an MCP server instance, including:
 * - Generating and persisting a unique instance identifier
 * - Tracking python-launched session processes
 * - Registering/unregistering the instance on startup/shutdown
 * - Providing the instance ID for Docker container labeling
 */
export class InstanceTracker {
  /** Unique UUID for this server instance. */
  readonly instanceId: string;
  /** Process ID of this server instance. */
  readonly pid: number;
  /** ISO 8601 timestamp of when this instance started. */
  readonly startedAt: string;
  /** Path to the instance metadata file. */
  readonly instanceFile: string;

  /** @internal Map from session name to PID */
  _pythonProcesses: Record<string, number>;

  /**
   * Initialize an InstanceTracker.
   *
   * This constructor should not be called directly. Use createAndRegister()
   * or loadFromFile() factory methods instead.
   *
   * @param instanceId - Unique UUID for this instance.
   * @param pid - Process ID of this server instance.
   * @param startedAt - ISO 8601 timestamp of when the instance started.
   */
  constructor(instanceId: string, pid: number, startedAt: string) {
    this.instanceId = instanceId;
    this.pid = pid;
    this.startedAt = startedAt;
    this._pythonProcesses = {};

    const instancesDir = path.join(os.homedir(), ".deephaven-mcp", "instances");
    this.instanceFile = path.join(instancesDir, `${instanceId}.json`);
  }

  /**
   * Create a new instance tracker and register it.
   *
   * This factory method creates a new instance with a unique UUID and immediately
   * persists it to disk. Call this on MCP server startup.
   *
   * @returns A new registered instance tracker.
   *
   * @example
   * ```typescript
   * const instance = await InstanceTracker.createAndRegister();
   * ```
   */
  static async createAndRegister(): Promise<InstanceTracker> {
    const instanceId = randomUUID();
    const pid = process.pid;
    const startedAt = new Date().toISOString();

    // Ensure instances directory exists
    const instancesDir = path.join(os.homedir(), ".deephaven-mcp", "instances");
    await fs.mkdir(instancesDir, { recursive: true });

    const tracker = new InstanceTracker(instanceId, pid, startedAt);
    await tracker._save();

    _logger.info(`[InstanceTracker] Registered new instance ${instanceId} (PID: ${pid})`);

    return tracker;
  }

  /**
   * Load an existing instance tracker from a metadata file.
   *
   * This is used during orphan cleanup to load information about other
   * (potentially dead) server instances.
   *
   * @param instanceFile - Path to the instance metadata JSON file.
   * @returns Instance tracker loaded from the file.
   * @throws If the instance file does not exist, contains invalid JSON, or is missing required fields.
   */
  static loadFromFile(instanceFile: string): InstanceTracker {
    // Note: this is intentionally synchronous to match the Python version
    const raw = require("node:fs").readFileSync(instanceFile, "utf-8");
    const data = JSON.parse(raw) as {
      instance_id: string;
      pid: number;
      started_at: string;
      python_processes?: Record<string, number>;
    };

    const tracker = new InstanceTracker(data["instance_id"], data["pid"], data["started_at"]);
    tracker._pythonProcesses = data["python_processes"] ?? {};

    return tracker;
  }

  /**
   * Track a new python-launched session process.
   *
   * Adds the process to the instance metadata so it can be cleaned up
   * if the server crashes or is killed.
   *
   * @param sessionName - Name of the session.
   * @param pid - Process ID of the python-launched deephaven-server process.
   *
   * @example
   * ```typescript
   * await instance.trackPythonProcess("my-session", 12345);
   * ```
   */
  async trackPythonProcess(sessionName: string, pid: number): Promise<void> {
    this._pythonProcesses[sessionName] = pid;
    await this._save();

    _logger.debug(
      `[InstanceTracker] Tracking python process for session '${sessionName}' (PID: ${pid})`,
    );
  }

  /**
   * Stop tracking a python-launched session process.
   *
   * Removes the process from the instance metadata, typically called when
   * the session is stopped normally. If the session name is not currently
   * tracked, this method is a no-op.
   *
   * @param sessionName - Name of the session to stop tracking.
   *
   * @example
   * ```typescript
   * await instance.untrackPythonProcess("my-session");
   * ```
   */
  async untrackPythonProcess(sessionName: string): Promise<void> {
    if (sessionName in this._pythonProcesses) {
      delete this._pythonProcesses[sessionName];
      await this._save();

      _logger.debug(`[InstanceTracker] Stopped tracking python process for session '${sessionName}'`);
    }
  }

  /**
   * Unregister this instance and remove its metadata file.
   *
   * Call this on normal server shutdown to clean up instance tracking.
   * This prevents the cleanup logic from attempting to clean up resources
   * for a server that shut down normally.
   *
   * @example
   * ```typescript
   * // In lifespan finally block
   * await instance.unregister();
   * ```
   */
  async unregister(): Promise<void> {
    try {
      await fs.unlink(this.instanceFile);
      _logger.info(`[InstanceTracker] Unregistered instance ${this.instanceId}`);
    } catch (e: unknown) {
      const err = e as NodeJS.ErrnoException;
      if (err.code !== "ENOENT") {
        _logger.warn(`[InstanceTracker] Error unregistering instance ${this.instanceId}: ${err.message}`);
      }
      // ENOENT means file already gone — treat as success (missing_ok=True)
    }
  }

  /**
   * Save instance metadata to disk using atomic write.
   *
   * Persists the current state of the instance tracker, including tracked
   * python processes, to the instance metadata file. Uses atomic write
   * (temp file + rename) to ensure the file is never left in a corrupted
   * state if the write is interrupted.
   */
  async _save(): Promise<void> {
    const data = {
      instance_id: this.instanceId,
      pid: this.pid,
      started_at: this.startedAt,
      python_processes: this._pythonProcesses,
    };

    const tempFile = this.instanceFile.replace(/\.json$/, ".tmp");
    await fs.writeFile(tempFile, JSON.stringify(data, null, 2), "utf-8");
    await fs.rename(tempFile, this.instanceFile);
  }
}

// ---------------------------------------------------------------------------
// isProcessRunning
// ---------------------------------------------------------------------------

/**
 * Check if a process with the given PID is currently running.
 *
 * Uses `process.kill(pid, 0)` which sends signal 0 (no-op) to probe
 * process existence without terminating it. Works on macOS and Linux;
 * on Windows, throws if the process does not exist.
 *
 * @param pid - Process ID to check.
 * @returns True if the process is running, False otherwise.
 *
 * Note:
 *   Returns false if the process does not exist. Unlike psutil.pid_exists(),
 *   this may return false due to permission errors on some platforms.
 */
export function isProcessRunning(pid: number): boolean {
  try {
    process.kill(pid, 0);
    return true;
  } catch {
    return false;
  }
}

// ---------------------------------------------------------------------------
// cleanupOrphanedResources
// ---------------------------------------------------------------------------

/**
 * Clean up orphaned Docker containers and python processes from dead server instances.
 *
 * This function should be called on MCP server startup. It:
 * 1. Scans the instances directory for registered server instances
 * 2. Checks if each instance's process is still running
 * 3. For dead instances, cleans up their orphaned resources:
 *    - Stops and removes Docker containers (via instance label)
 *    - Terminates python processes (from instance metadata)
 * 4. Removes instance metadata files for dead instances
 *
 * The cleanup is safe for concurrent server instances - only resources from
 * dead servers are cleaned up. Running servers are left untouched.
 *
 * @example
 * ```typescript
 * // In lifespan, before yielding context
 * await cleanupOrphanedResources();
 * ```
 *
 * Note:
 *   Errors during cleanup are logged but don't raise exceptions, ensuring
 *   that server startup continues even if cleanup partially fails.
 */
export async function cleanupOrphanedResources(
  instancesDir: string = path.join(os.homedir(), ".deephaven-mcp", "instances"),
): Promise<void> {

  let dirEntries: string[];
  try {
    dirEntries = await fs.readdir(instancesDir);
  } catch (e: unknown) {
    const err = e as NodeJS.ErrnoException;
    if (err.code === "ENOENT") {
      _logger.debug("[InstanceTracker] No instances directory, skipping orphan cleanup");
      return;
    }
    throw e;
  }

  const instanceFiles = dirEntries
    .filter((f) => f.endsWith(".json"))
    .map((f) => path.join(instancesDir, f));

  if (instanceFiles.length === 0) {
    _logger.debug("[InstanceTracker] No instance files found, skipping orphan cleanup");
    return;
  }

  _logger.info(
    `[InstanceTracker] Checking ${instanceFiles.length} instance(s) for orphaned resources...`,
  );

  for (const instanceFile of instanceFiles) {
    try {
      const tracker = InstanceTracker.loadFromFile(instanceFile);

      if (isProcessRunning(tracker.pid)) {
        _logger.debug(
          `[InstanceTracker] Instance ${tracker.instanceId} still running (PID ${tracker.pid}), skipping`,
        );
        continue;
      }

      _logger.warn(
        `[InstanceTracker] Found dead instance ${tracker.instanceId} (PID ${tracker.pid}), cleaning up orphans...`,
      );

      await _cleanupDockerContainersForInstance(tracker.instanceId);
      await _cleanupPythonProcessesForInstance(tracker);

      try {
        await fs.unlink(instanceFile);
      } catch {
        // ignore if already gone
      }

      _logger.info(
        `[InstanceTracker] Cleaned up orphaned resources for instance ${tracker.instanceId}`,
      );
    } catch (e: unknown) {
      const err = e as Error;
      _logger.error(
        `[InstanceTracker] Error cleaning up instance ${path.basename(instanceFile)}: ${err.message}`,
      );
    }
  }
}

/**
 * Test-seam alias for `cleanupOrphanedResources` that accepts an explicit instances directory.
 *
 * Because `vi.spyOn(os, "homedir")` cannot work in ESM (Node built-in namespaces are
 * not configurable), tests use this function directly with a temp directory.
 *
 * @param instancesDir - Absolute path to the instances directory to scan.
 */
export const _cleanupInstancesDir = cleanupOrphanedResources;

// ---------------------------------------------------------------------------
// _cleanupDockerContainersForInstance
// ---------------------------------------------------------------------------

/**
 * Clean up Docker containers for a specific server instance.
 *
 * Finds all Docker containers labeled with the instance ID and stops/removes them.
 * Uses the 'deephaven-mcp-server-instance' label to identify containers belonging
 * to the dead instance.
 *
 * @param instanceId - The instance UUID to clean up containers for.
 *
 * Note:
 *   Errors during cleanup are logged but do not raise exceptions.
 */
export async function _cleanupDockerContainersForInstance(instanceId: string): Promise<void> {
  try {
    const stdout = await _runCommand("docker", [
      "ps",
      "-a",
      "--filter",
      `label=deephaven-mcp-server-instance=${instanceId}`,
      "--format",
      "{{.ID}}",
    ]);

    if (stdout === null) {
      return;
    }

    const containerIds = stdout
      .trim()
      .split("\n")
      .map((s) => s.trim())
      .filter((s) => s.length > 0);

    if (containerIds.length === 0) {
      _logger.debug(`[InstanceTracker] No Docker containers found for instance ${instanceId}`);
      return;
    }

    _logger.info(
      `[InstanceTracker] Found ${containerIds.length} orphaned container(s) for instance ${instanceId}`,
    );

    for (const containerId of containerIds) {
      _logger.info(`[InstanceTracker] Stopping orphaned container ${containerId.slice(0, 12)}...`);

      // Try to stop gracefully (ignore errors — container may already be stopped)
      await _runCommand("docker", ["stop", containerId]);

      // Remove the container
      await _runCommand("docker", ["rm", containerId]);

      _logger.info(`[InstanceTracker] Cleaned up orphaned container ${containerId.slice(0, 12)}`);
    }
  } catch (e: unknown) {
    const err = e as Error;
    _logger.error(
      `[InstanceTracker] Error cleaning up Docker containers for instance ${instanceId}: ${err.message}`,
    );
  }
}

// ---------------------------------------------------------------------------
// _cleanupPythonProcessesForInstance
// ---------------------------------------------------------------------------

/**
 * Clean up python processes for a specific server instance.
 *
 * Terminates all python processes tracked in the instance metadata. Sends
 * SIGTERM to each tracked process after verifying it's still running.
 * Processes that are already dead are logged and skipped.
 *
 * @param tracker - The instance tracker with python process information.
 *
 * Note:
 *   Errors during process termination (e.g., permission denied, process already
 *   exited) are logged as warnings but do not raise exceptions.
 */
export async function _cleanupPythonProcessesForInstance(tracker: InstanceTracker): Promise<void> {
  const pythonProcesses = tracker._pythonProcesses;

  if (Object.keys(pythonProcesses).length === 0) {
    _logger.debug(
      `[InstanceTracker] No python processes found for instance ${tracker.instanceId}`,
    );
    return;
  }

  _logger.info(
    `[InstanceTracker] Found ${Object.keys(pythonProcesses).length} orphaned python process(es) for instance ${tracker.instanceId}`,
  );

  for (const [sessionName, pid] of Object.entries(pythonProcesses)) {
    try {
      if (isProcessRunning(pid)) {
        _logger.info(
          `[InstanceTracker] Terminating orphaned python process ${pid} (session: ${sessionName})`,
        );
        process.kill(pid, "SIGTERM");
      } else {
        _logger.debug(
          `[InstanceTracker] Python process ${pid} (session: ${sessionName}) already dead`,
        );
      }
    } catch (e: unknown) {
      const err = e as Error;
      _logger.warn(
        `[InstanceTracker] Error terminating python process ${pid} (session: ${sessionName}): ${err.message}`,
      );
    }
  }
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/**
 * Run a subprocess command and return its stdout, or null on failure.
 *
 * @param cmd - Command to run.
 * @param args - Arguments for the command.
 * @returns Stdout string on success, null if the command fails.
 */
async function _runCommand(cmd: string, args: string[]): Promise<string | null> {
  return new Promise<string | null>((resolve) => {
    const child = spawn(cmd, args, { stdio: ["ignore", "pipe", "pipe"] });
    let stdout = "";
    let stderr = "";

    child.stdout.on("data", (chunk: Buffer) => {
      stdout += chunk.toString();
    });
    child.stderr.on("data", (chunk: Buffer) => {
      stderr += chunk.toString();
    });

    child.on("close", (code) => {
      if (code !== 0) {
        _logger.debug(`[InstanceTracker] Command ${cmd} ${args.join(" ")} failed (code ${code}): ${stderr.trim()}`);
        resolve(null);
      } else {
        resolve(stdout);
      }
    });

    child.on("error", (err) => {
      _logger.debug(`[InstanceTracker] Command ${cmd} failed to start: ${err.message}`);
      resolve(null);
    });
  });
}
